### PR TITLE
cmd/tsconnect,util/precompress: move precompression to its own package

### PR DESCRIPTION
**Pull request by mihaip**
We have very similar code in corp, moving it to tsweb allows it to be
reused.

Updates #5133

Signed-off-by: Mihai Parparita <mihai@tailscale.com>

### CODE REVIEW

File: tsweb/precompress.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	return eg.Wait()
+}
+
+type PrecompressOptions struct {
+	// FastCompression controls whether compression should be optimized for
+	// speed rather than size.
+	FastCompression bool
+	// ProgressFn is invoked when a file in the directory is about to be
+	// compressed.
+	ProgressFn func(path string)
```
Comment (bradfitz) [inline tsweb/precompress.go:59]: is this optional?

File: tsweb/precompress.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/andybalholm/brotli"
+	"golang.org/x/sync/errgroup"
+)
+
+// PrecompressDir compresses static assets in dirPath using Gzip and Brotli, so
+// that they can be later served with OpenPrecompressedFile. Meant to be run
```
Comment (bradfitz) [inline tsweb/precompress.go:22]: It's meant ....

File: tsweb/precompress.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+// during build steps, to avoid the cost of repeatedly compressing unchanging
+// resources.
+func PrecompressDir(dirPath string, options PrecompressOptions) error {
```
Comment (bradfitz) [inline tsweb/precompress.go:25]: kinda weird to have this in the tsweb package.

It also means all our tsweb code now brings in brotli?

Comment (mihaip) [reply tsweb/precompress.go:25]: I can move it to `util/precompress`?

File: tsweb/precompress.go (hunk 1 above)
Comment (mihaip) [reply tsweb/precompress.go]: Yes. Added a comment, is there a canonical way to indicate that it is?

Comment (mihaip) [reply tsweb/precompress.go]: Looks like a "if non-nil" clause is idiomatic.

File: tsweb/precompress.go (hunk 3 above)
Comment (mihaip) [reply tsweb/precompress.go]: Moved to `util/precompress` as discussed.

**Review by bradfitz — APPROVED** [review]

**Pull request merged**